File: src/domain/transfers/stateMachine.ts
import { ApiError } from '../../errors';
import type { TransferStatus } from '../../contract';

/**
 * Allowed transitions. A transition not listed here is a bug or a race — the
 * engine and any ops action must go through assertTransition().
 *
 *   CREATED → QUOTED → SCREENED → AWAITING_FUNDS → FUNDED → SETTLING
 *           → SETTLED → PAYING_OUT → COMPLETED
 *
 * plus EXPIRED (pre-funding timeout), REJECTED (screening / insufficient
 * funds / partner failure), and REVERSING → REVERSED (post-settlement unwind).
 */
export const ALLOWED_TRANSITIONS: Record<TransferStatus, readonly TransferStatus[]> = {
  CREATED: ['QUOTED', 'EXPIRED'],
  QUOTED: ['SCREENED', 'EXPIRED'],
  SCREENED: ['AWAITING_FUNDS', 'REJECTED', 'EXPIRED'],
  AWAITING_FUNDS: ['FUNDED', 'REJECTED', 'EXPIRED'],
  FUNDED: ['SETTLING', 'REJECTED'],
  SETTLING: ['SETTLED', 'REJECTED'],
  SETTLED: ['PAYING_OUT'],
  PAYING_OUT: ['COMPLETED'],
  COMPLETED: ['REVERSING'],
  REJECTED: [],
  EXPIRED: [],
  REVERSING: ['REVERSED'],
  REVERSED: [],
};

/** The single forward step on the happy path, or null at a waypoint/terminal. */
export const FORWARD_STEP: Partial<Record<TransferStatus, TransferStatus>> = {
  CREATED: 'QUOTED',
  QUOTED: 'SCREENED',
  SCREENED: 'AWAITING_FUNDS',
  AWAITING_FUNDS: 'FUNDED',
  FUNDED: 'SETTLING',
  SETTLING: 'SETTLED',
  SETTLED: 'PAYING_OUT',
  PAYING_OUT: 'COMPLETED',
};

export function canTransition(from: TransferStatus, to: TransferStatus): boolean {
  return ALLOWED_TRANSITIONS[from].includes(to);
}

export function assertTransition(from: TransferStatus, to: TransferStatus): void {
  if (!canTransition(from, to)) {
    throw new ApiError('CONFLICT', `Transfer cannot move from ${from} to ${to}.`, false);
  }
}

File: src/domain/transfers/state_machine.rs
//! Explicit transition table. Every state change — engine or ops — goes
//! through `assert_transition`.

use crate::contract::transfer::TransferStatus::{self, *};
use crate::error::ApiError;

pub fn allowed(from: TransferStatus) -> &'static [TransferStatus] {
    match from {
        Created => &[Quoted, Expired],
        Quoted => &[Screened, Expired],
        Screened => &[AwaitingFunds, Rejected, Expired],
        AwaitingFunds => &[Funded, Rejected, Expired],
        Funded => &[Settling, Rejected],
        Settling => &[Settled, Rejected],
        Settled => &[PayingOut],
        PayingOut => &[Completed],
        Completed => &[Reversing],
        Rejected => &[],
        Expired => &[],
        Reversing => &[Reversed],
        Reversed => &[],
    }
}

/// The single forward step on the happy path, or None at a waypoint/terminal.
pub fn forward_step(from: TransferStatus) -> Option<TransferStatus> {
    Some(match from {
        Created => Quoted,
        Quoted => Screened,
        Screened => AwaitingFunds,
        AwaitingFunds => Funded,
        Funded => Settling,
        Settling => Settled,
        Settled => PayingOut,
        PayingOut => Completed,
        _ => return None,
    })
}

pub fn assert_transition(from: TransferStatus, to: TransferStatus) -> Result<(), ApiError> {
    if allowed(from).contains(&to) {
        Ok(())
    } else {
        Err(ApiError::conflict(format!(
            "Transfer cannot move from {} to {}.",
            from.as_str(),
            to.as_str()
        )))
    }
}
